### PR TITLE
Support export to storage account set in config

### PR DESCRIPTION
**Pull request by namadabu**
## Description
For bulk export, we require users to pass in a _destinationType and _destinationConnectionSettings parameter along with the request. This information will be used to connect to the appropriate destination and export data there. With the PR, we are removing this functionality and changing the way export works. The storage account information should now be added to the app config and this will be used for exporting data. The default template has also been updated to support such a scenario.

TODO:
-Export Destination client should handle token expiry
-Add more tests to ExportJobTask

## Related issues
Addresses [issue [AB#69774](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/69774) [AB#71879](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/71879) [AB#71880](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/71880)].

## Testing
Existing tests. Manually tested by deploying a fhir-server instance (using the updated template) and verified that I could export data to the associated storage account.

### CODE REVIEW

File: samples/templates/default-azuredeploy.json [hunk 1]
```diff
@@ -158,7 +167,10 @@
             "FhirServer:Security:Authentication:Authority": "[parameters('securityAuthenticationAuthority')]",
             "FhirServer:Security:Authentication:Audience": "[parameters('securityAuthenticationAudience')]",
             "CosmosDb:ContinuationTokenSizeLimitInKb": "1",
-            "DataStore": "CosmosDb"
+            "DataStore": "CosmosDb",
```
Comment (brendankowitz) [inline samples/templates/default-azuredeploy.json]: What happens with the SQL template deployment?

Comment (namadabu) [reply samples/templates/default-azuredeploy.json]: I haven't added these changes this to the SQL template since we don't yet support export for SQL.

File: samples/templates/default-azuredeploy.json [hunk 2]
```diff
@@ -126,14 +126,22 @@
...
+            "defaultValue": "R4",
             "allowedValues": [
                 "Stu3",
                 "R4"
             ],
             "metadata": {
                 "description": "Only applies when specifying Repository Url"
             }
+        },
+        "storageAccountName": {
```
Comment (jackliums) [inline samples/templates/default-azuredeploy.json]: Should we derive the storage account name instead of asking user to enter it so it behaves like other dependencies (e.g., KeyVault, AppInsights and etc)?

I added the storage account integration for the dicom-server template: https://github.com/microsoft/dicom-server/pull/23/files

File: samples/templates/default-azuredeploy.json [hunk 3]
```diff
@@ -340,6 +352,34 @@
...
+            "type": "Microsoft.Storage/storageAccounts/providers/roleAssignments",
+            "apiVersion": "2018-09-01-preview",
+            "name": "[concat(parameters('storageAccountName'), '/Microsoft.Authorization/', guid(uniqueString(parameters('storageAccountName'))))]",
+            "dependsOn": [
+                "[parameters('storageAccountName')]",
+                "[variables('serviceName')]"
+            ],
+            "properties": {
+                "roleDefinitionId": "[variables('roleDefinitionId')]",
+                "principalId": "[reference(concat('Microsoft.Web/sites/', variables('serviceName')), '2018-11-01', 'full').identity.principalId]"
```
Comment (jackliums) [inline samples/templates/default-azuredeploy.json]: Should minimize to only permission needed (e.g., blob)

File: src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public string AccessTokenProviderType => "azure";
+
+        public async Task<string> GetAccessTokenForResourceAsync(Uri resourceUri, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(resourceUri, nameof(resourceUri));
+
+            string accessToken = await _azureServiceTokenProvider.GetAccessTokenAsync(resourceUri.ToString(), cancellationToken: cancellationToken);
+            if (accessToken == null)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs]: should it throw exception if it fails to get the access token instead of null?

File: src/Microsoft.Health.Fhir.Core/Configs/ExportJobConfiguration.cs [hunk 5]
```diff
@@ -15,6 +15,23 @@ public class ExportJobConfiguration
         /// </summary>
         public bool Enabled { get; set; }
 
+        /// <summary>
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Configs/ExportJobConfiguration.cs]: This will be a breaking change so we need to be explicit about how we want to handle it.

If this code is deployed and there is an existing export job record, the new code will fail, which is fine if that's what we decide to do. Just want to make sure we think about breaking change scenarios when changing existing code.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AccessTokenProvider/InMemoryAccessTokenProvider.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export.AccessTokenProvider
+{
+    public class InMemoryAccessTokenProvider : IAccessTokenProvider
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AccessTokenProvider/InMemoryAccessTokenProvider.cs]: Is this needed in the production code?

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/Models/DestinationInfo.cs [hunk 7]
```diff
@@ -12,9 +11,6 @@ public class DestinationInfo
     {
         public DestinationInfo(string destinationType, string destinationConnectionString)
         {
-            EnsureArg.IsNotNullOrWhiteSpace(destinationType, nameof(destinationType));
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/Models/DestinationInfo.cs]: This class is no longer needed.

File: src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs [hunk 8]
```diff
@@ -28,24 +24,8 @@ internal class ValidateExportRequestFilterAttribute : ActionFilterAttribute
...
-        public ValidateExportRequestFilterAttribute(IExportDestinationClientFactory exportDestinationClientFactory)
+        public ValidateExportRequestFilterAttribute()
         {
-            EnsureArg.IsNotNull(exportDestinationClientFactory, nameof(exportDestinationClientFactory));
-
-            _exportDestinationClientFactory = exportDestinationClientFactory;
-
-            _supportedQueryParams = new HashSet<string>(StringComparer.Ordinal)
-            {
-                KnownQueryParameterNames.DestinationType,
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs]: Remove these two from KnownQueryParameterNames

File: src/Microsoft.Health.Fhir.R4.Web/appsettings.json [hunk 9]
```diff
@@ -38,7 +38,10 @@
         },
         "Operations": {
             "Export": {
-                "Enabled": false,
+                "Enabled": true,
+                "StorageAccountType": null,
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.R4.Web/appsettings.json]: Can this be move to top level? E.g., a storage account that can be used for other purpose too not just export?

File: src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            _logger = logger;
+        }
+
+        public string AccessTokenProviderType => "azure";
+
+        public async Task<string> GetAccessTokenForResourceAsync(Uri resourceUri, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(resourceUri, nameof(resourceUri));
+
+            string accessToken = await _azureServiceTokenProvider.GetAccessTokenAsync(resourceUri.ToString(), cancellationToken: cancellationToken);
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs]: Are there any kind of transient errors it would make sense to have a retry here?  Or do we retry at a higher level?

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs [hunk 11]
```diff
@@ -211,16 +208,33 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
             }
         }
 
-        // Get destination info from secret store, create appropriate export client and connect to destination.
+        // Get destination information, create appropriate export client and connect to destination.
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs]: I don't think we want to aquire token on every loop right? The token is valid for certain period of time, so we should be able to reuse it until we have to refresh it?

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportDestinationClient/IExportDestinationClient.cs [hunk 12]
```diff
@@ -29,6 +29,16 @@ public interface IExportDestinationClient
         /// <returns>A <see cref="Task"/> representing connection operation.</returns>
         Task ConnectAsync(string connectionSettings, CancellationToken cancellationToken, string containerId = null);
 
+        /// <summary>
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportDestinationClient/IExportDestinationClient.cs]: Have you considered maybe delegate the "authentication" part and glue them together in IoC container?

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs (hunk 11 above)
Comment (jackliums) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs]: Also, i don't think it's the export job task's responsibility to acquire access token.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobConfigurationValidator.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        public bool ValidateExportJobConfig()
+        {
+            if (string.IsNullOrWhiteSpace(_exportJobConfiguration.StorageAccountType) ||
+                !_exportDestinationClientFactory.IsSupportedDestinationType(_exportJobConfiguration.StorageAccountType))
+            {
+                throw new ExportJobConfigValidationException(string.Format(Resources.UnsupportedDestinationType, _exportJobConfiguration.StorageAccountType), HttpStatusCode.BadRequest);
+            }
+
+            // Check whether the config contains a uri to a storage account or a connection string.
+            if (Uri.TryCreate(_exportJobConfiguration.StorageAccountConnection, UriKind.Absolute, out Uri resultUri))
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobConfigurationValidator.cs]: You already have an interface for handling acess token (i would make it more generic so it can handle how authentication is handled) and then just separate the connection to another implementation of the interface rather than depending on whther the conection is a "url" or not.

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs [hunk 14]
```diff
@@ -78,16 +77,14 @@ public class ExportController : Controller
...
-            [FromQuery(Name = KnownQueryParameterNames.DestinationConnectionSettings)] string destinationConnectionString)
+        public async Task<IActionResult> Export()
         {
             if (!_exportConfig.Enabled)
             {
                 throw new RequestNotValidException(string.Format(Resources.UnsupportedOperation, OperationsConstants.Export));
             }
 
-            CreateExportResponse response = await _mediator.ExportAsync(_fhirRequestContextAccessor.FhirRequestContext.Uri, destinationType, destinationConnectionString, HttpContext.RequestAborted);
+            CreateExportResponse response = await _mediator.ExportAsync(_fhirRequestContextAccessor.FhirRequestContext.Uri, HttpContext.RequestAborted);
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs]: A few more things we should address (doesn't necessarily need to be in this PR).

1. If the user cancels the export or if export fails, we should delete all exported data from the storage.
2. We need to figure out how to expire exported results for completed job.

File: src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs (hunk 8 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs]: Removed.

File: samples/templates/default-azuredeploy.json (hunk 2 above)
Comment (namadabu) [reply samples/templates/default-azuredeploy.json]: Update the template to generate storage account name. Also added an option for users to choose whether they want to enable export.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/Models/DestinationInfo.cs (hunk 7 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/Models/DestinationInfo.cs]: Deleted

File: src/Microsoft.Health.Fhir.R4.Web/appsettings.json (hunk 9 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.R4.Web/appsettings.json]: I feel that it would be okay to keep it within export for now and move it later when we decide to use it for other purposes.

File: src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs (hunk 4 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs]: Changed to throw an exception when we don't get back an access token

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureExportDestinationClient.cs [hunk 15]
```diff
@@ -42,26 +55,89 @@ public async Task ConnectAsync(string connectionSettings, CancellationToken canc
...
             if (!CloudStorageAccount.TryParse(decodedConnectionString, out CloudStorageAccount cloudAccount))
             {
                 throw new DestinationConnectionException(Resources.InvalidConnectionSettings, HttpStatusCode.BadRequest);
             }
 
             _blobClient = cloudAccount.CreateCloudBlobClient();
+            await CreateContainerAsync(_blobClient, containerId);
+        }
+
+        public async Task ConnectHelperAsync(string connectionString, CancellationToken cancellationToken, string containerId = null)
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureExportDestinationClient.cs]: Renaming and merging this with `ConnectWithAccessTokenAsync()`

File: samples/templates/default-azuredeploy.json (hunk 1 above)
Comment (namadabu) [reply samples/templates/default-azuredeploy.json]: Update the sql template also with export options since we know support export in SQL also!

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AccessTokenProvider/InMemoryAccessTokenProvider.cs (hunk 6 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AccessTokenProvider/InMemoryAccessTokenProvider.cs]: Removed `InMemoryAccessTokenProvider`.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs (hunk 11 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs]: The original change had it only getting the token at the beginning of the export process. Once we connect, we get into the while loop for processing the export. Like you mentioned, I have moved the  acquiring access token logic out of the `ExportJobTask` class. Now that logic is abstracted away as part of the `IExportDestinationClient`

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportDestinationClient/IExportDestinationClient.cs (hunk 12 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportDestinationClient/IExportDestinationClient.cs]: Removed the `ConnectWithAccessTokenAsync` method

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs (hunk 14 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs]: TTL on the document could be a solution for this (but would need a different solution for sql)

File: src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs (hunk 10 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Azure/AzureAccessTokenProvider.cs]: Will handle retries as part of a separate PR since this PR is already huge. Created WI #72260 to track it

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs (hunk 14 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs]: Created work items to track these issues (and other nice to haves)

File: samples/templates/default-azuredeploy.json (hunk 3 above)
Comment (namadabu) [reply samples/templates/default-azuredeploy.json]: The `roleDefinitionId` variable is defined to be the guid that corresponds to the `Storage Blob Data Contributor` role (Reference - https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles). If you are referring to the 'full' in the `reference` call, that just corresponds to returning the full object (so that we can get the `identity.principalId` of the app service.

File: src/Microsoft.Health.Fhir.Core/Microsoft.Health.Fhir.Core.csproj [hunk 16]
```diff
@@ -25,6 +25,7 @@
     <PackageReference Include="Ensure.That" Version="8.1.2" />
     <PackageReference Include="FluentValidation" Version="8.6.1" />
     <PackageReference Include="MediatR" Version="7.0.0" />
+    <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" Version="1.3.1" />
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Core/Microsoft.Health.Fhir.Core.csproj]: Removed this

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobConfigurationValidator.cs (hunk 13 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobConfigurationValidator.cs]: Refactored the code to handle authentication in a separate interface/class and removed the `ExportJobConfigurationValidator`

File: samples/templates/default-azuredeploy-sql.json [hunk 17]
```diff
@@ -138,6 +145,8 @@
         "msdeployPackageUrlToUse": "[if(empty(parameters('msdeployPackageUrl')),variables('defaultMsdeployPackageUrl'),parameters('msdeployPackageUrl'))]",
         "deployAppInsights": "[and(parameters('deployApplicationInsights'),not(variables('isMAG')))]",
         "appInsightsName": "[concat('AppInsights-', variables('serviceName'))]",
+        "roleDefinitionId": "[concat('/subscriptions/', subscription().subscriptionId, '/providers/Microsoft.Authorization/roleDefinitions/', 'ba92f5b4-2d11-453d-a403-e96b0029c9fe')]",
```
Comment (jackliums) [inline samples/templates/default-azuredeploy-sql.json]: nit: perhaps a more specific name? like storageBlobDataContributerRoleId?

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureConnectionStringClientInitializer.cs [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+            {
+                throw new ExportClientInitializerException(Resources.InvalidConnectionSettings, HttpStatusCode.BadRequest);
+            }
+
+            if (!CloudStorageAccount.TryParse(_exportJobConfiguration.StorageAccountConnection, out CloudStorageAccount cloudAccount))
+            {
+                throw new ExportClientInitializerException(Resources.InvalidConnectionSettings, HttpStatusCode.BadRequest);
+            }
+
+            CloudBlobClient blobClient = cloudAccount.CreateCloudBlobClient();
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureConnectionStringClientInitializer.cs]: if this method fails to parse the connection string, would it throw exception? if so, does it need to be translated to ExportClientInitializerException?

File: src/Microsoft.Health.Fhir.R4.Web/appsettings.json [hunk 19]
```diff
@@ -39,7 +39,10 @@
...
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.R4.Web/appsettings.json]: You could simplify even further by remove this and only support blob storage in the Azure implementation.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs [hunk 20]
```diff
@@ -211,16 +188,18 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
             }
         }
 
-        // Get destination info from secret store, create appropriate export client and connect to destination.
-        private async Task GetDestinationInfoAndConnectAsync(CancellationToken cancellationToken)
+        // Create appropriate export client and connect to destination.
+        private async Task GetDestinationClientAndConnectAsync(CancellationToken cancellationToken)
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs]: This still doesn't address the issue with expiring access token (e.g., it's acquiring token before the while loop but the while loop could take longer than the validation period of the access token). Would that be fixed in this PR or another PR?

File: src/Microsoft.Health.Fhir.Core/Configs/ExportJobConfiguration.cs [hunk 21]
```diff
@@ -15,6 +15,25 @@ public class ExportJobConfiguration
...
+        /// Should be a connection string to the required storage account.
+        /// </summary>
+        public string StorageAccountConnection { get; set; }
+
+        /// <summary>
+        /// Determines the storage account connection that will be used to export data to.
+        /// Should be a uri pointing to the required storage account.
+        /// </summary>
+#pragma warning disable CA1056 // Uri properties should not be strings
+        public string StorageAccountUri { get; set; }
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Configs/ExportJobConfiguration.cs]: Have you tried to make this Uri instead of string?

File: samples/templates/default-azuredeploy-sql.json (hunk 17 above)
Comment (namadabu) [reply samples/templates/default-azuredeploy-sql.json]: Updated the name

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs (hunk 20 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs]: Yes. I will add that ability along with retries as part of another PR. The corresponding WI is #72260

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureConnectionStringClientInitializer.cs (hunk 18 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureConnectionStringClientInitializer.cs]: The method description does not mention anything about throwing an exception. And I assumed it should be okay to call this since we successfully parse the connection string in the previous step. To be safe, I have added a try-catch block around this call also.

File: src/Microsoft.Health.Fhir.R4.Web/appsettings.json (hunk 19 above)
Comment (namadabu) [reply src/Microsoft.Health.Fhir.R4.Web/appsettings.json]: Deleted this config and related classes that depended on this flow. Also moved the `InMemoryExportDestinationClient` to the tests project
